project-loader/src/main/kotlin/Args.kt

```diff
@@ -12,8 +12,19 @@ private fun <T> splitAndCreate(str: String, creator: (String, String) -> T): T {
     return creator(split[0], split[1])
 }
 
+private fun <T> splitAndCreatePlugin(str: String, creator: (String, String, Boolean) -> T): T {
+    val split = str.split("::")
+    if (split.size < 2) {
+        throw RuntimeException("string is not of the right format. Expected <key>::<value>[::<pre installed>]")
+    }
+    if (split.size == 3){
+        return creator(split[0], split[1], split[2].toBoolean())
+    }
+    return creator(split[0], split[1], false)
```

Comment:
With your suggestions we could also get rid of this line 23 and the `if(split.size == 3)` statement I guess.

Reply:
This depends how you would like to approach the complete solution. What you changed so far is the internal data structures of the project loader. These data structures are used only by the project loader and aren't what the user puts into their gradle files. 

The project loader is part of a command line application that runs the desired code inside of MPS. The architecture of the mps gradles plugin never runs MPS inside the JVM that is actually executing the gradle build. The mps gradle plugin starts a new JVM/process for all of it MPS related code. The project loader is responsible for setting up the MPS environment and then delegates execution to the action that should be performed on the project, e.g. the model check. The project loader has a command line interface to configure the MPS environment incl. the project to open. 

What you changes so far is the command line interface but not the part that is what user uses in their gradle files. The part that is responsible for the gradle side is in `ModelcheckMpsProjectPlugin`. You can see there that is passes command line arguments to the new JVM its starting which also include the plugins. The class `BasePluginExtensions` contains all the common arguments to configure the MPS environment. In fact it's essentially the very same as in the `Args` class but tailored for use inside of gralde scripts. 

The parsing code with the split etc. that you changed is used parse the command line arguments. Keeping that backward compatible is fine and makes a lot of sense because not every is using gradle as a front end to start the project loader. There are for instance usecases where the project loader / modelchecker are called from maven. 

The next step would be to change the `BasePluginExtensions` to include the "preinstalled" flag and adapt `argsFromBaseExtension` to include it into the command line arguments.

Reply:
```suggestion
    return creator(split[0], split[1], true)
```